Add migration cli

paul-sachs opened the pull request:
Created a CLI that migrates from old packages to new.

- Using jscodeshift to guarantee that we are only modifying imports and not arbitrary code
- Testing with jest since jscodeshift have a bunch of helpers to use with jest. We can migrate those to jasmine for consistency

timostamm commented on packages/connect-migrate/src/cli.ts:
```diff
@@ -0,0 +1,98 @@
+  } else if (existsSync(path.join(dir, "yarn.lock"))) {
+    packageManager = "yarn";
+  }
+  assert(
+    packageManager !== undefined,
+    "Could not determine package manager based on lock file. Make sure you have a lock file (yarn.lock, package-lock.json, or pnpm-lock.yaml) in your project.",
+  );
```
We cannot use assertions for flow control. It's easy to run the command in the wrong directory, and we have to consider it to be part of the regular flow, not an exceptional case.

In general, command line tools have to:
- exit with status 0 on success
- only crash with stack trace on an exceptional error
- print a short error message (no stack traces) and usage information on stderr and exit with non-zero code for invalid arguments
- print a short error message and exit with non-zero code in case of an error
- print usage information for the flag `-h` or `--help`
- print the version for the flag `--version` (or `-v` unless taken for verbose output)

Here is an example for parsing a command line:

```ts
interface CommandLineArgs {
  ok: true;
  help: boolean;
  version: boolean;
}

interface CommandLineError {
  ok: false;
  errorMessage: string;
}

export function parseCommandLineArgs(args: string[]): CommandLineArgs | CommandLineError {
  const parsed = {
    version: false,
    help: false
  };
  while (args.length > 0) {
    const arg = args.shift();
    switch (arg) {
      case "-h":
      case "--help":
        if (parsed.help) {
          break;
        }
        parsed.help = true;
        break;
      case "--version":
        parsed.version = true;
        break;
      default:
        return {
          ok: false,
          errorMessage: `unknown argument ${arg}`
        };
    }
  }
  return {...parsed, ok: true};
}
```

Used as:

```ts
import { parseCommandLineArgs } from "./arguments";
import process from "node:process";

const args = parseCommandLineArgs(process.argv.slice(2));
if (!args.ok) {
  process.stderr.write(`${args.errorMessage}\n`);
  process.exit(1);
}
if (args.version) {
  // todo: use version from package.json
  process.stdout.write("connect-migrate v0.1.0\n");
  process.exit(0);
}
if (args.help) {
  process.stdout.write(`USAGE: connect-migrate\n`);
  process.stdout.write(`Updates references to connect-es packages in your project to use @connectrpc.\n`);
  process.exit(0);
}

// todo: logic
process.exit(0);
```

This scales up to everything we'll need here, and makes testing trivial.

timostamm commented on packages/connect-migrate/src/replacement-map.ts:
```diff
@@ -0,0 +1,51 @@
+const keys = Object.keys(replacementMap) as (keyof typeof replacementMap)[];
+
+export function getReplacementImport(sourceValue: string): string | undefined {
+  const match = keys.find((key) => {
+    return sourceValue.match(new RegExp(`^${key}\\b`));
+  });
+  if (match !== undefined) {
+    return sourceValue.replace(
+      new RegExp(`^${match}\\b`),
+      replacementMap[match],
+    );
+  }
```
Constructing regexp dynamically is fragile 😅

Can't we simply do?

```suggestion
  if (sourceValue.startsWith("@bufbuild/connect")) {
    return "@connectrpc/connnect" + sourceValue.substring("@bufbuild/connect".length);
  }
  return undefined;
```

paul-sachs replied:
And this solution also prevents us from accidentally modifying packages we don't want to. I guess I just wanted to be very specific about which packages I replace, and the provided map of replacement packages makes that very explicit.

paul-sachs replied:
Actually, found a different way to do this without relying on dynamic regexp. Updated.

timostamm commented on packages/connect-migrate/src/replacement-map.ts:
```diff
@@ -0,0 +1,51 @@
+export function replacePackageJSONReferences(jsonString: string): string {
+  let result = jsonString;
+  for (const key of keys) {
+    result = result.replace(
+      new RegExp(`"${key}":`, "g"),
+      `"${replacementMap[key]}":`,
+    );
```
Let's avoid constructing regex dynamically 😅 

WDYT about

```suggestion
    result = result.replace(
      "${key}", `"${replacementMap[key]}"`,
    );
```

? 

Deliberately omitted the colon, it's valid JSON to have a space between.

paul-sachs replied:
ah good call.

timostamm commented on packages/connect-migrate/src/transforms/__testfixtures__/modify-imports/sub-directory-imports.input.ts:
```diff
@@ -0,0 +1,3 @@
+import { makeAnyClient } from "@bufbuild/connect/dist/cjs/any-client";
```
Not a valid import. Let's use

```suggestion
import { runUnaryCall } from "@bufbuild/connect/protocol";
```

paul-sachs replied:
yeah at this point I was just tossing some random cases and not considering valid imports. I can modify

timostamm commented on packages/connect-migrate/src/transforms/__testfixtures__/modify-imports/simple.input.ts:
```diff
@@ -0,0 +1,4 @@
+import { createPromiseClient } from "@bufbuild/connect";
```
Single quotes will be very common. Are they preserved?

paul-sachs replied:
Just pushed some code up to detect the style used (based on the first import detected) and we maintain that style.

timostamm commented on packages/connect-migrate/src/cli.ts:
```diff
@@ -0,0 +1,98 @@
+  }
+  console.log("Updated package dependencies.");
+  console.log("Updating references...");
+  executeInContext(jscodeshiftExecutable, [
```
Does jscodeshift have a programmatic API? It would be nice to avoid spawning a process...

paul-sachs replied:
It does, we can combine that directly with fastGlob and should be able to use that.

timostamm reviewed (commented):
This looks very promising, now we just have to sort out the details. 

Blocking on jest, we already have enough dependencies. I'm okay with replacing it later, when we know better what needs testing.

paul-sachs commented:
@timostamm all great points, i had started on this tool separate from connect-es, so i'll migrate the packages/testing to something more akin to what's in connectrpc already. This was mostly just porting things over as is. I'll tune it up.

paul-sachs commented on packages/connect-migrate/src/cli.ts:
```diff
@@ -0,0 +1,98 @@
+  console.log("Updated package dependencies.");
+  console.log("Updating references...");
+  executeInContext(jscodeshiftExecutable, [
+    "--ignore-pattern=**/node_modules/**",
```
Thankfully the output of jscodeshift is pretty thorough so it'll be obvious what files have changed. I'll check to make sure that if we call it programatically, we still get that output.

paul-sachs replied:
Added a section to the readme to indicate which files are changed. That probably should suffice for now.

paul-sachs commented on packages/connect-migrate/src/replacement-map.ts:
```diff
@@ -0,0 +1,51 @@
+  return undefined;
+}
+
+export function replacePackageJSONReferences(jsonString: string): string {
```
Hmm good call, I'll add a warning if install fails.

timostamm commented on packages/connect-migrate/src/transforms/modify-imports.spec.ts:
```diff
@@ -0,0 +1,57 @@
+import transform from "./modify-imports";
+
+// Make sure to import the testfixtures from source so we're not comparing transformed fixtures
+const fixtureLocation = join(
+  __dirname,
+  "../../../src/transforms",
+  "__testfixtures__",
+);
```
I looked into ways to run codeshift the same way in tests and in the CLI, but the workers do give us a significant bump in performance for large numbers of files.

Since we can't run the same code paths in tests as we do in the CLI, we can just as well remove the `__testfixtures__` directory and work with strings here in the tests. 

The we end up with something like this:

```ts
import jscodeshift from "jscodeshift";
import transform from "./modify-imports.js";

function t(source: string) {
  const shift = jscodeshift.withParser("tsx");
  return transform(
    { path: "test-file", source },
    {
      jscodeshift: shift,
      j: shift,
      stats: () => {},
      report: () => {},
    },
    {},
  );
}

describe("modify-imports", () => {
  it("should modify import", () => {
    const got = `import a from "@bufbuild/connect";`;
    const want = `import a from "@connectrpc/connect";`;
    expect(t(got)?.trim()).toBe(want.trim());
  });
  it("should modify import with single quotes", () => {
    const got = `import a from '@bufbuild/connect';`;
    const want = `import a from '@connectrpc/connect';`;
    expect(t(got)?.trim()).toBe(want.trim());
  });
  it("should modify subpath import", () => {
    const got = `import a from "@bufbuild/connect/protocol";`;
    const want = `import a from "@connectrpc/connect/protocol";`;
    expect(t(got)?.trim()).toBe(want.trim());
  });
  it("should not modify irrelevant import", () => {
    const got = `import a from "@foobar/connect";`;
    const want = `import a from "@foobar/connect";`;
    expect(t(got)?.trim()).toBe(want.trim());
  });
  it("should modify require", () => {
    const got = `const a = require("@bufbuild/connect");`;
    const want = `const a = require("@connectrpc/connect");`;
    expect(t(got)?.trim()).toBe(want.trim());
  });
  it("should not modify irrelevant require", () => {
    const got = `const a = require("@foobar/connect");`;
    const want = `const a = require("@foobar/connect");`;
    expect(t(got)?.trim()).toBe(want.trim());
  });
  it("should modify require with single quotes", () => {
    const got = `const a = require('@bufbuild/connect');`;
    const want = `const a = require('@connectrpc/connect');`;
    expect(t(got)?.trim()).toBe(want.trim());
  });
});
```

I believe we don't lose any confidence with these tests (it's clear now that the file name is irrelevant), but we can remove a lot of files and ignores.

What do you think?

paul-sachs replied:
Yeah that sounds fine to me.

timostamm replied:
Adding it in https://github.com/connectrpc/connect-es/pull/774

paul-sachs commented on packages/connect-migrate/src/cli.ts:
```diff
@@ -0,0 +1,132 @@
+void main();
+
+function reinstallDependencies(dir: string) {
+  // TODO: Perhaps we can find all the lock files and run install wherever we find them. This'll help for repos without a single
+  // top level lock file.
```
Yeah actually working on this right now, mostly to ease migration of examples.

paul-sachs commented on packages/connect-migrate/src/arguments.ts:
```diff
@@ -0,0 +1,57 @@
+Updates references to connect-es packages in your project to use @connectrpc.
+`;
+
+export function parseCommandLineArgs(
```
Yeah, alright, I liked having the help colocated with the argument parsing, will adjust.

paul-sachs commented:
@timostamm I've fixed the issue with parsing the issue you had with the `custom-transport.ts` file and we now use different parsers for different filetypes. This does raise the risk that someone might have some custom babel setup but i've added a section to the readme to describe other ways to run our transforms.
